### PR TITLE
Migrate network-setup to nftables and improve it into a better state

### PR DESCRIPTION
**Pull request by kanpov**
## Changes

network-for-clones is not changed in this PR since it's a whole another can of worms I'd like to deal with later.
These changes are to network-setup:

- Migrate to nftables
- Improve the documentation itself to convert it from a rather basic getting started guide to something more usable in the real world:

- [x] Multiple guests section
- [x] Guest kernel configuration at kernel level section (`ip` kernel boot arg)

## Reason

Fulfills #4874 

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this
  PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet
  [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

______________________________________________________________________

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by kanpov** [comment]
I'm welcome to critique of the new sections in the doc, since I think this is the real minimal amount of information needed for a Firecracker user to get a proper networking setup with support for multiple guests and other improvements. Back when I started working with Firecracker, I found this specific area of the docs to be **especially** frustrating to use:

- Why shouldn't I use it if I just have another iface instead of eth0, why not just say you can replace the name?
- Why no indications on how to support multiple guests?
- Why no explanations of what the iptables or ip commands actually do for people unfamiliar with NAT, TUN/TAP and gateways?

With this, I've tried to address these.

**Comment by kanpov** [comment]
> Thanks for your contribution! I need to try the commands myself, but looks good so far. 
> 
> I have a couple of asks: could we squash the commits together? I think it becomes easier to review. And we would like to keep the `iptables` commands for some time until `nft` becomes more spread.

I'll apply the squash and, as for iptables, I like your idea for collapsible blocks as opt-in for using it.

 Iptables, however, is most certainly no longer the default and, even when it's used, it's used only through the iptables-nft compat layer which is already itself deprecated on some distros (including RHEL).

File: docs/network-setup.md [hunk 1]
```diff
@@ -107,7 +138,121 @@ your environment. For testing, you can add a public DNS server to
...
+other rule is 2:
+```bash
+sudo nft delete rule firecracker postrouting handle 1
+sudo nft delete rule firecracker filter handle 2
+```
+
+_Advanced:_ If you created a bridge interface, delete it using the following:
+```bash
+sudo ip link del br0
+```
```
[the remark this answers is not in the document]
Comment (kanpov) [reply docs/network-setup.md:173]: This is unchanged from the current version, but sure.

**Comment by kanpov** [comment]
@pb8o Sorry for the delay! I've fixed the problems, and made an effort to support both `nft` and `iptables-nft` seamlessly. For the methods of routing, I explicitly highlighted 3 of them: NAT, referring to the main portion of the article, bridge-based, referring to the dedicated section, and namespaced, referring to "Network for Clones" (which will need another PR to perform some extremely needed fixing as well).

I explicitly don't want to encourage using the legacy `iptables` command instead of `iptables-nft`, since it leads to both `nf_tables` and `x_tables` kernel routing modules being used at the same time and them conflicting with each other while trying to both do packet filtering each according to their own rules. So, I suggest `iptables-nft` in the current revision, which has the same syntax so should be compatible with your requirements, but doesn't cause such conflicts by internally translating commands to `nft` syntax.

**Comment by kanpov** [comment]
For the squashing, I'm pretty sure the PR can just be merged with the "squash and merge" option to clean everything up.

**Comment by pb8o** [comment]
> I explicitly don't want to encourage using the legacy `iptables` command instead of `iptables-nft`

In my Ubuntu machine the default is `iptables-nft`, but I guess it doesn't hurt to make it explicit.

```
% ll /usr/sbin/iptables
lrwxrwxrwx 1 root root 26 2024-02-22T09:25 /usr/sbin/iptables -> /etc/alternatives/iptables
% ll /etc/alternatives/iptables
lrwxrwxrwx 1 root root 22 2024-02-22T09:23 /etc/alternatives/iptables -> /usr/sbin/iptables-nft
% iptables --version
iptables v1.8.7 (nf_tables)
```

**Review by pb8o — COMMENTED** [review]
Thanks, this looks good to me!

**Comment by pb8o** [comment]
Ok so I saw some style checks failing, how to fix:

1. Run `./tools/devtool fmt` and `mdformat` will reformat the markdown to its liking. 
2. Could you squash the commits into one? We use `gitlint` and make sure each commit follows the [contribution guidelines](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md)

You can check if the styles check pass by running `./tools/devtool checkstyle`

**Comment by kanpov** [comment]
Messed up the commit tree, will have to remake the PR(

**Pull request closed without merging**